File: apps/api/internal/http/dto/airport_intelligence.go
package dto

import (
	"strings"
	"time"

	"github.com/AsifAbbasov/global-flight-analytics/apps/api/internal/domain/airport"

	"github.com/AsifAbbasov/global-flight-analytics/apps/api/internal/airportintelligence/airportproduction"
	"github.com/AsifAbbasov/global-flight-analytics/apps/api/internal/airportintelligence/passport"
	"github.com/AsifAbbasov/global-flight-analytics/apps/api/internal/airportintelligence/ranking"
	"github.com/AsifAbbasov/global-flight-analytics/apps/api/internal/airportintelligence/statistics"
	"github.com/AsifAbbasov/global-flight-analytics/apps/api/internal/airportintelligence/trends"
)

type AirportIntelligenceWindow struct {
	StartTime     time.Time `json:"start_time"`
	EndTime       time.Time `json:"end_time"`
	AsOfTime      time.Time `json:"as_of_time"`
	CompletedDays int       `json:"completed_days"`
}
type AirportIntelligenceLimitation struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}
type AirportPassportIdentity struct {
	ICAOCode string `json:"icao_code"`
	IATACode string `json:"iata_code"`
	Name     string `json:"name"`
}
type AirportPassportLocation struct {
	City            string                  `json:"city"`
	Country         string                  `json:"country"`
	Latitude        float64                 `json:"latitude"`
	Longitude       float64                 `json:"longitude"`
	ElevationM      *float64                `json:"elevation_m"`
	ElevationStatus airport.ElevationStatus `json:"elevation_status"`
	Timezone        string                  `json:"timezone"`
}
type AirportPassportOperations struct {
	Arrivals       int `json:"arrivals"`
	Departures     int `json:"departures"`
	Activity       int `json:"activity"`
	ActiveAircraft int `json:"active_aircraft"`
}
type AirportPassportDataQuality struct {
	FreshnessScore float64   `json:"freshness_score"`
	CoverageScore  float64   `json:"coverage_score"`
	ObservedAt     time.Time `json:"observed_at"`
}
type AirportPassportResponse struct {
	Identity    AirportPassportIdentity    `json:"identity"`
	Location    AirportPassportLocation    `json:"location"`
	Operations  AirportPassportOperations  `json:"operations"`
	DataQuality AirportPassportDataQuality `json:"data_quality"`
	Description string                     `json:"description"`
	GeneratedAt time.Time                  `json:"generated_at"`
}
type AirportStatisticsResponse struct {
	ICAOCode            string    `json:"icao_code"`
	WindowStart         time.Time `json:"window_start"`
	WindowEnd           time.Time `json:"window_end"`
	Arrivals            int       `json:"arrivals"`
	Departures          int       `json:"departures"`
	TotalMovements      int       `json:"total_movements"`
	ArrivalShare        float64   `json:"arrival_share"`
	DepartureShare      float64   `json:"departure_share"`
	MovementsPerHour    float64   `json:"movements_per_hour"`
	ActiveAircraft      int       `json:"active_aircraft"`
	ActiveRoutes        int       `json:"active_routes"`
	ObservedSamples     int       `json:"observed_samples"`
	ExpectedSamples     int       `json:"expected_samples"`
	CoverageScore       float64   `json:"coverage_score"`
	FreshnessScore      float64   `json:"freshness_score"`
	LatestObservationAt time.Time `json:"latest_observation_at"`
	GeneratedAt         time.Time `json:"generated_at"`
}
type AirportRankingSummaryResponse struct {
	Position              int     `json:"position"`
	TotalAirports         int     `json:"total_airports"`
	ActivityScore         float64 `json:"activity_score"`
	DataConfidence        float64 `json:"data_confidence"`
	MovementsComponent    float64 `json:"movements_component"`
	RoutesComponent       float64 `json:"routes_component"`
	ObservationsComponent float64 `json:"observations_component"`
	IntensityComponent    float64 `json:"intensity_component"`
}
type AirportIntelligenceOverviewResponse struct {
	Version     string                          `json:"version"`
	Window      AirportIntelligenceWindow       `json:"window"`
	Passport    AirportPassportResponse         `json:"passport"`
	Statistics  AirportStatisticsResponse       `json:"statistics"`
	Ranking     AirportRankingSummaryResponse   `json:"ranking"`
	Limitations []AirportIntelligenceLimitation `json:"limitations"`
	GeneratedAt time.Time                       `json:"generated_at"`
}
type AirportIntelligenceHistoryResponse struct {
	Version     string                          `json:"version"`
	Window      AirportIntelligenceWindow       `json:"window"`
	ICAOCode    string                          `json:"icao_code"`
	Entries     []AirportStatisticsResponse     `json:"entries"`
	Limitations []AirportIntelligenceLimitation `json:"limitations"`
	GeneratedAt time.Time                       `json:"generated_at"`
}
type AirportTrendPointResponse struct {
	WindowStart      time.Time `json:"window_start"`
	WindowEnd        time.Time `json:"window_end"`
	TotalMovements   int       `json:"total_movements"`
	MovementsPerHour float64   `json:"movements_per_hour"`
	ActiveRoutes     int       `json:"active_routes"`
	CoverageScore    float64   `json:"coverage_score"`
	FreshnessScore   float64   `json:"freshness_score"`
}
type AirportIntelligenceTrendsResponse struct {
	Version                            string                          `json:"version"`
	Window                             AirportIntelligenceWindow       `json:"window"`
	ICAOCode                           string                          `json:"icao_code"`
	ComparedWindows                    int                             `json:"compared_windows"`
	WindowDurationSeconds              int64                           `json:"window_duration_seconds"`
	Direction                          string                          `json:"direction"`
	Baseline                           AirportTrendPointResponse       `json:"baseline"`
	Current                            AirportTrendPointResponse       `json:"current"`
	Peak                               AirportTrendPointResponse       `json:"peak"`
	TotalMovementsChange               int                             `json:"total_movements_change"`
	MovementsPerHourChange             float64                         `json:"movements_per_hour_change"`
	MovementsPerHourChangePercent      float64                         `json:"movements_per_hour_change_percent"`
	MovementsPerHourChangePercentKnown bool                            `json:"movements_per_hour_change_percent_known"`
	ActiveRoutesChange                 int                             `json:"active_routes_change"`
	CoverageScoreChange                float64                         `json:"coverage_score_change"`
	FreshnessScoreChange               float64                         `json:"freshness_score_change"`
	GapCount                           int                             `json:"gap_count"`
	GapDurationSeconds                 int64                           `json:"gap_duration_seconds"`
	ObservedDurationSeconds            int64                           `json:"observed_duration_seconds"`
	ContinuityScore                    float64                         `json:"continuity_score"`
	Limitations                        []AirportIntelligenceLimitation `json:"limitations"`
	GeneratedAt                        time.Time                       `json:"generated_at"`
}
type AirportRankedItemResponse struct {
	Position              int     `json:"position"`
	ICAOCode              string  `json:"icao_code"`
	IATACode              string  `json:"iata_code"`
	Name                  string  `json:"name"`
	City                  string  `json:"city"`
	Country               string  `json:"country"`
	ActivityScore         float64 `json:"activity_score"`
	DataConfidence        float64 `json:"data_confidence"`
	MovementsComponent    float64 `json:"movements_component"`
	RoutesComponent       float64 `json:"routes_component"`
	ObservationsComponent float64 `json:"observations_component"`
	IntensityComponent    float64 `json:"intensity_component"`
	CoverageScore         float64 `json:"coverage_score"`
	FreshnessScore        float64 `json:"freshness_score"`
	TotalMovements        int     `json:"total_movements"`
	ActiveRoutes          int     `json:"active_routes"`
	ObservedSamples       int     `json:"observed_samples"`
	ExpectedSamples       int     `json:"expected_samples"`
	MovementsPerHour      float64 `json:"movements_per_hour"`
	ActiveAircraft        int     `json:"active_aircraft"`
}
type AirportRankingWeightsResponse struct {
	Movements    float64 `json:"movements"`
	Routes       float64 `json:"routes"`
	Observations float64 `json:"observations"`
	Intensity    float64 `json:"intensity"`
	Coverage     float64 `json:"coverage"`
	Freshness    float64 `json:"freshness"`
}
type AirportIntelligenceRankingResponse struct {
	Version     string                          `json:"version"`
	Window      AirportIntelligenceWindow       `json:"window"`
	Weights     AirportRankingWeightsResponse   `json:"weights"`
	Airports    []AirportRankedItemResponse     `json:"airports"`
	Limitations []AirportIntelligenceLimitation `json:"limitations"`
	GeneratedAt time.Time                       `json:"generated_at"`
}

func ToAirportIntelligenceOverviewResponse(result airportproduction.OverviewResult) AirportIntelligenceOverviewResponse {
	value := result.Overview
	return AirportIntelligenceOverviewResponse{Version: result.Version, Window: toAirportIntelligenceWindow(result.Window), Passport: toAirportPassportResponse(value.Passport), Statistics: toAirportStatisticsResponse(value.Statistics), Ranking: AirportRankingSummaryResponse{Position: value.Ranking.Position, TotalAirports: value.Ranking.TotalAirports, ActivityScore: value.Ranking.ActivityScore, DataConfidence: value.Ranking.DataConfidence, MovementsComponent: value.Ranking.MovementsComponent, RoutesComponent: value.Ranking.RoutesComponent, ObservationsComponent: value.Ranking.ObservationsComponent, IntensityComponent: value.Ranking.IntensityComponent}, Limitations: toAirportIntelligenceLimitations(result.Limitations), GeneratedAt: result.GeneratedAt}
}
func ToAirportIntelligenceHistoryResponse(result airportproduction.HistoryResult) AirportIntelligenceHistoryResponse {
	entries := make([]AirportStatisticsResponse, 0, len(result.History.Entries))
	for _, entry := range result.History.Entries {
		entries = append(entries, toAirportStatisticsResponse(entry))
	}
	return AirportIntelligenceHistoryResponse{Version: result.Version, Window: toAirportIntelligenceWindow(result.Window), ICAOCode: result.History.ICAOCode, Entries: entries, Limitations: toAirportIntelligenceLimitations(result.Limitations), GeneratedAt: result.GeneratedAt}
}
func ToAirportIntelligenceTrendsResponse(result airportproduction.TrendsResult) AirportIntelligenceTrendsResponse {
	value := result.Trends
	return AirportIntelligenceTrendsResponse{Version: result.Version, Window: toAirportIntelligenceWindow(result.Window), ICAOCode: value.ICAOCode, ComparedWindows: value.ComparedWindows, WindowDurationSeconds: int64(value.WindowDuration / time.Second), Direction: string(value.Direction), Baseline: toAirportTrendPointResponse(value.Baseline), Current: toAirportTrendPointResponse(value.Current), Peak: toAirportTrendPointResponse(value.Peak), TotalMovementsChange: value.TotalMovementsChange, MovementsPerHourChange: value.MovementsPerHourChange, MovementsPerHourChangePercent: value.MovementsPerHourChangePercent, MovementsPerHourChangePercentKnown: value.MovementsPerHourChangePercentKnown, ActiveRoutesChange: value.ActiveRoutesChange, CoverageScoreChange: value.CoverageScoreChange, FreshnessScoreChange: value.FreshnessScoreChange, GapCount: value.GapCount, GapDurationSeconds: int64(value.GapDuration / time.Second), ObservedDurationSeconds: int64(value.ObservedDuration / time.Second), ContinuityScore: value.ContinuityScore, Limitations: toAirportIntelligenceLimitations(result.Limitations), GeneratedAt: result.GeneratedAt}
}
func ToAirportIntelligenceRankingResponse(result airportproduction.RankingResult, limit int) AirportIntelligenceRankingResponse {
	ranked := result.Ranking.Airports
	if limit > 0 && len(ranked) > limit {
		ranked = ranked[:limit]
	}
	items := make([]AirportRankedItemResponse, 0, len(ranked))
	for _, item := range ranked {
		airportValue := result.Airports[strings.ToUpper(strings.TrimSpace(item.ICAOCode))]
		items = append(items, toAirportRankedItemResponse(item, airportValue.IATACode, airportValue.Name, airportValue.City, airportValue.Country))
	}
	return AirportIntelligenceRankingResponse{Version: result.Version, Window: toAirportIntelligenceWindow(result.Window), Weights: AirportRankingWeightsResponse{Movements: result.Ranking.ActivityWeights.Movements, Routes: result.Ranking.ActivityWeights.Routes, Observations: result.Ranking.ActivityWeights.Observations, Intensity: result.Ranking.ActivityWeights.Intensity, Coverage: result.Ranking.ConfidenceWeights.Coverage, Freshness: result.Ranking.ConfidenceWeights.Freshness}, Airports: items, Limitations: toAirportIntelligenceLimitations(result.Limitations), GeneratedAt: result.GeneratedAt}
}
func toAirportIntelligenceWindow(value airportproduction.Window) AirportIntelligenceWindow {
	return AirportIntelligenceWindow{StartTime: value.StartTime, EndTime: value.EndTime, AsOfTime: value.AsOfTime, CompletedDays: value.CompletedDays}
}
func toAirportIntelligenceLimitations(values []airportproduction.Limitation) []AirportIntelligenceLimitation {
	result := make([]AirportIntelligenceLimitation, 0, len(values))
	for _, value := range values {
		result = append(result, AirportIntelligenceLimitation{Code: value.Code, Message: value.Message})
	}
	return result
}
func toAirportPassportResponse(value passport.Passport) AirportPassportResponse {
	elevationM, elevationStatus := ToAirportElevation(
		value.Location.ElevationM,
		value.Location.ElevationAvailable,
	)
	return AirportPassportResponse{Identity: AirportPassportIdentity{ICAOCode: value.Identity.ICAOCode, IATACode: value.Identity.IATACode, Name: value.Identity.Name}, Location: AirportPassportLocation{City: value.Location.City, Country: value.Location.Country, Latitude: value.Location.Latitude, Longitude: value.Location.Longitude, ElevationM: elevationM, ElevationStatus: elevationStatus, Timezone: value.Location.Timezone}, Operations: AirportPassportOperations{Arrivals: value.Operations.Arrivals, Departures: value.Operations.Departures, Activity: value.Operations.Activity, ActiveAircraft: value.Operations.ActiveAircraft}, DataQuality: AirportPassportDataQuality{FreshnessScore: value.DataQuality.FreshnessScore, CoverageScore: value.DataQuality.CoverageScore, ObservedAt: value.DataQuality.ObservedAt}, Description: value.Description, GeneratedAt: value.GeneratedAt}
}
func toAirportStatisticsResponse(value statistics.Statistics) AirportStatisticsResponse {
	return AirportStatisticsResponse{ICAOCode: value.ICAOCode, WindowStart: value.WindowStart, WindowEnd: value.WindowEnd, Arrivals: value.Arrivals, Departures: value.Departures, TotalMovements: value.TotalMovements, ArrivalShare: value.ArrivalShare, DepartureShare: value.DepartureShare, MovementsPerHour: value.MovementsPerHour, ActiveAircraft: value.ActiveAircraft, ActiveRoutes: value.ActiveRoutes, ObservedSamples: value.ObservedSamples, ExpectedSamples: value.ExpectedSamples, CoverageScore: value.CoverageScore, FreshnessScore: value.FreshnessScore, LatestObservationAt: value.LatestObservationAt, GeneratedAt: value.GeneratedAt}
}
func toAirportTrendPointResponse(value trends.Point) AirportTrendPointResponse {
	return AirportTrendPointResponse{WindowStart: value.WindowStart, WindowEnd: value.WindowEnd, TotalMovements: value.TotalMovements, MovementsPerHour: value.MovementsPerHour, ActiveRoutes: value.ActiveRoutes, CoverageScore: value.CoverageScore, FreshnessScore: value.FreshnessScore}
}
func toAirportRankedItemResponse(value ranking.AirportRank, iataCode, name, city, country string) AirportRankedItemResponse {
	return AirportRankedItemResponse{Position: value.Position, ICAOCode: value.ICAOCode, IATACode: iataCode, Name: name, City: city, Country: country, ActivityScore: value.ActivityScore, DataConfidence: value.DataConfidence, MovementsComponent: value.MovementsComponent, RoutesComponent: value.RoutesComponent, ObservationsComponent: value.ObservationsComponent, IntensityComponent: value.IntensityComponent, CoverageScore: value.CoverageScore, FreshnessScore: value.FreshnessScore, TotalMovements: value.TotalMovements, ActiveRoutes: value.ActiveRoutes, ObservedSamples: value.ObservedSamples, ExpectedSamples: value.ExpectedSamples, MovementsPerHour: value.MovementsPerHour, ActiveAircraft: value.ActiveAircraft}
}

File: apps/web/lib/api/airport-intelligence.ts
// FRONTEND_UNIFIED_AIRPORT_ANALYTICS_WORKSPACE_V1

import {
  APIRequestError,
  requestAPIData,
} from '@/lib/api/client'
import type {
  AirportIntelligenceHistory,
  AirportIntelligenceLimitation,
  AirportIntelligenceOverview,
  AirportIntelligenceRanking,
  AirportIntelligenceRankingOptions,
  AirportIntelligenceTrends,
  AirportIntelligenceWindow,
  AirportIntelligenceWindowOptions,
  AirportPassport,
  AirportRankedItem,
  AirportRankingSummary,
  AirportRankingWeights,
  AirportStatistics,
  AirportTrendPoint,
} from '@/types/airport-intelligence'

const minimumDays = 1
const maximumDays = 365
const minimumLimit = 1
const maximumLimit = 200

export async function getAirportIntelligenceRanking(
  options: AirportIntelligenceRankingOptions
): Promise<AirportIntelligenceRanking> {
  const searchParams = buildSearchParams(options)
  searchParams.set('limit', String(normalizeInteger(options.limit, 'limit', minimumLimit, maximumLimit)))

  return parseRanking(
    await requestAPIData<unknown>('/api/v1/airports/intelligence/ranking', {
      signal: options.signal,
      searchParams,
    })
  )
}

export async function getAirportIntelligenceOverview(
  icaoCode: string,
  options: AirportIntelligenceWindowOptions
): Promise<AirportIntelligenceOverview> {
  return parseOverview(
    await requestAPIData<unknown>(
      `/api/v1/airports/${encodeURIComponent(normalizeICAOCode(icaoCode))}/intelligence/overview`,
      {
        signal: options.signal,
        searchParams: buildSearchParams(options),
      }
    )
  )
}

export async function getAirportIntelligenceHistory(
  icaoCode: string,
  options: AirportIntelligenceWindowOptions
): Promise<AirportIntelligenceHistory> {
  return parseHistory(
    await requestAPIData<unknown>(
      `/api/v1/airports/${encodeURIComponent(normalizeICAOCode(icaoCode))}/intelligence/history`,
      {
        signal: options.signal,
        searchParams: buildSearchParams(options),
      }
    )
  )
}

export async function getAirportIntelligenceTrends(
  icaoCode: string,
  options: AirportIntelligenceWindowOptions
): Promise<AirportIntelligenceTrends> {
  return parseTrends(
    await requestAPIData<unknown>(
      `/api/v1/airports/${encodeURIComponent(normalizeICAOCode(icaoCode))}/intelligence/trends`,
      {
        signal: options.signal,
        searchParams: buildSearchParams(options),
      }
    )
  )
}

function buildSearchParams(options: AirportIntelligenceWindowOptions): URLSearchParams {
  const searchParams = new URLSearchParams()
  searchParams.set('days', String(normalizeInteger(options.days, 'days', minimumDays, maximumDays)))
  return searchParams
}

function normalizeICAOCode(value: string): string {
  const normalized = value.trim().toUpperCase()
  if (!/^[A-Z0-9]{4}$/.test(normalized)) {
    throw new APIRequestError('Airport ICAO code must contain exactly four letters or digits.')
  }
  return normalized
}

function normalizeInteger(
  value: number,
  field: string,
  minimum: number,
  maximum: number
): number {
  if (!Number.isInteger(value) || value < minimum || value > maximum) {
    throw new APIRequestError(
      `Airport Intelligence ${field} must be between ${minimum} and ${maximum}.`
    )
  }
  return value
}

function parseRanking(value: unknown): AirportIntelligenceRanking {
  const item = objectValue(value, 'ranking')
  return {
    version: stringValue(item.version, 'ranking.version'),
    window: parseWindow(item.window, 'ranking.window'),
    weights: parseWeights(item.weights, 'ranking.weights'),
    airports: arrayValue(item.airports, 'ranking.airports').map((entry, index) =>
      parseRankedItem(entry, `ranking.airports[${index}]`)
    ),
    limitations: parseLimitations(item.limitations, 'ranking.limitations'),
    generated_at: timestampValue(item.generated_at, 'ranking.generated_at'),
  }
}

function parseOverview(value: unknown): AirportIntelligenceOverview {
  const item = objectValue(value, 'overview')
  return {
    version: stringValue(item.version, 'overview.version'),
    window: parseWindow(item.window, 'overview.window'),
    passport: parsePassport(item.passport, 'overview.passport'),
    statistics: parseStatistics(item.statistics, 'overview.statistics'),
    ranking: parseRankingSummary(item.ranking, 'overview.ranking'),
    limitations: parseLimitations(item.limitations, 'overview.limitations'),
    generated_at: timestampValue(item.generated_at, 'overview.generated_at'),
  }
}

function parseHistory(value: unknown): AirportIntelligenceHistory {
  const item = objectValue(value, 'history')
  return {
    version: stringValue(item.version, 'history.version'),
    window: parseWindow(item.window, 'history.window'),
    icao_code: icaoValue(item.icao_code, 'history.icao_code'),
    entries: arrayValue(item.entries, 'history.entries').map((entry, index) =>
      parseStatistics(entry, `history.entries[${index}]`)
    ),
    limitations: parseLimitations(item.limitations, 'history.limitations'),
    generated_at: timestampValue(item.generated_at, 'history.generated_at'),
  }
}

function parseTrends(value: unknown): AirportIntelligenceTrends {
  const item = objectValue(value, 'trends')
  return {
    version: stringValue(item.version, 'trends.version'),
    window: parseWindow(item.window, 'trends.window'),
    icao_code: icaoValue(item.icao_code, 'trends.icao_code'),
    compared_windows: integerValue(item.compared_windows, 'trends.compared_windows'),
    window_duration_seconds: integerValue(
      item.window_duration_seconds,
      'trends.window_duration_seconds'
    ),
    direction: stringValue(item.direction, 'trends.direction'),
    baseline: parseTrendPoint(item.baseline, 'trends.baseline'),
    current: parseTrendPoint(item.current, 'trends.current'),
    peak: parseTrendPoint(item.peak, 'trends.peak'),
    total_movements_change: integerValue(
      item.total_movements_change,
      'trends.total_movements_change',
      true
    ),
    movements_per_hour_change: finiteNumber(
      item.movements_per_hour_change,
      'trends.movements_per_hour_change'
    ),
    movements_per_hour_change_percent: finiteNumber(
      item.movements_per_hour_change_percent,
      'trends.movements_per_hour_change_percent'
    ),
    movements_per_hour_change_percent_known: booleanValue(
      item.movements_per_hour_change_percent_known,
      'trends.movements_per_hour_change_percent_known'
    ),
    active_routes_change: integerValue(
      item.active_routes_change,
      'trends.active_routes_change',
      true
    ),
    coverage_score_change: finiteNumber(
      item.coverage_score_change,
      'trends.coverage_score_change'
    ),
    freshness_score_change: finiteNumber(
      item.freshness_score_change,
      'trends.freshness_score_change'
    ),
    gap_count: integerValue(item.gap_count, 'trends.gap_count'),
    gap_duration_seconds: integerValue(
      item.gap_duration_seconds,
      'trends.gap_duration_seconds'
    ),
    observed_duration_seconds: integerValue(
      item.observed_duration_seconds,
      'trends.observed_duration_seconds'
    ),
    continuity_score: ratioValue(item.continuity_score, 'trends.continuity_score'),
    limitations: parseLimitations(item.limitations, 'trends.limitations'),
    generated_at: timestampValue(item.generated_at, 'trends.generated_at'),
  }
}

function parseWindow(value: unknown, field: string): AirportIntelligenceWindow {
  const item = objectValue(value, field)
  return {
    start_time: timestampValue(item.start_time, `${field}.start_time`),
    end_time: timestampValue(item.end_time, `${field}.end_time`),
    as_of_time: timestampValue(item.as_of_time, `${field}.as_of_time`),
    completed_days: integerValue(item.completed_days, `${field}.completed_days`),
  }
}

function parseWeights(value: unknown, field: string): AirportRankingWeights {
  const item = objectValue(value, field)
  return {
    movements: ratioValue(item.movements, `${field}.movements`),
    routes: ratioValue(item.routes, `${field}.routes`),
    observations: ratioValue(item.observations, `${field}.observations`),
    intensity: ratioValue(item.intensity, `${field}.intensity`),
    coverage: ratioValue(item.coverage, `${field}.coverage`),
    freshness: ratioValue(item.freshness, `${field}.freshness`),
  }
}

function parseRankedItem(value: unknown, field: string): AirportRankedItem {
  const item = objectValue(value, field)
  return {
    position: integerValue(item.position, `${field}.position`),
    icao_code: icaoValue(item.icao_code, `${field}.icao_code`),
    iata_code: stringValue(item.iata_code, `${field}.iata_code`, true),
    name: stringValue(item.name, `${field}.name`),
    city: stringValue(item.city, `${field}.city`, true),
    country: stringValue(item.country, `${field}.country`, true),
    activity_score: ratioValue(item.activity_score, `${field}.activity_score`),
    data_confidence: ratioValue(item.data_confidence, `${field}.data_confidence`),
    movements_component: ratioValue(item.movements_component, `${field}.movements_component`),
    routes_component: ratioValue(item.routes_component, `${field}.routes_component`),
    observations_component: ratioValue(
      item.observations_component,
      `${field}.observations_component`
    ),
    intensity_component: ratioValue(item.intensity_component, `${field}.intensity_component`),
    coverage_score: ratioValue(item.coverage_score, `${field}.coverage_score`),
    freshness_score: ratioValue(item.freshness_score, `${field}.freshness_score`),
    total_movements: integerValue(item.total_movements, `${field}.total_movements`),
    active_routes: integerValue(item.active_routes, `${field}.active_routes`),
    observed_samples: integerValue(item.observed_samples, `${field}.observed_samples`),
    expected_samples: integerValue(item.expected_samples, `${field}.expected_samples`),
    movements_per_hour: nonNegativeNumber(
      item.movements_per_hour,
      `${field}.movements_per_hour`
    ),
    active_aircraft: integerValue(item.active_aircraft, `${field}.active_aircraft`),
  }
}

function parsePassport(value: unknown, field: string): AirportPassport {
  const item = objectValue(value, field)
  const identity = objectValue(item.identity, `${field}.identity`)
  const location = objectValue(item.location, `${field}.location`)
  const operations = objectValue(item.operations, `${field}.operations`)
  const quality = objectValue(item.data_quality, `${field}.data_quality`)
  return {
    identity: {
      icao_code: icaoValue(identity.icao_code, `${field}.identity.icao_code`),
      iata_code: stringValue(identity.iata_code, `${field}.identity.iata_code`, true),
      name: stringValue(identity.name, `${field}.identity.name`),
    },
    location: {
      city: stringValue(location.city, `${field}.location.city`, true),
      country: stringValue(location.country, `${field}.location.country`, true),
      latitude: boundedNumber(location.latitude, `${field}.location.latitude`, -90, 90),
      longitude: boundedNumber(location.longitude, `${field}.location.longitude`, -180, 180),
      elevation_m: nullableFiniteNumber(location.elevation_m, `${field}.location.elevation_m`),
      elevation_status: stringValue(
        location.elevation_status,
        `${field}.location.elevation_status`
      ),
      timezone: stringValue(location.timezone, `${field}.location.timezone`, true),
    },
    operations: {
      arrivals: integerValue(operations.arrivals, `${field}.operations.arrivals`),
      departures: integerValue(operations.departures, `${field}.operations.departures`),
      activity: integerValue(operations.activity, `${field}.operations.activity`),
      active_aircraft: integerValue(
        operations.active_aircraft,
        `${field}.operations.active_aircraft`
      ),
    },
    data_quality: {
      freshness_score: ratioValue(
        quality.freshness_score,
        `${field}.data_quality.freshness_score`
      ),
      coverage_score: ratioValue(
        quality.coverage_score,
        `${field}.data_quality.coverage_score`
      ),
      observed_at: timestampValue(
        quality.observed_at,
        `${field}.data_quality.observed_at`
      ),
    },
    description: stringValue(item.description, `${field}.description`, true),
    generated_at: timestampValue(item.generated_at, `${field}.generated_at`),
  }
}

function parseStatistics(value: unknown, field: string): AirportStatistics {
  const item = objectValue(value, field)
  return {
    icao_code: icaoValue(item.icao_code, `${field}.icao_code`),
    window_start: timestampValue(item.window_start, `${field}.window_start`),
    window_end: timestampValue(item.window_end, `${field}.window_end`),
    arrivals: integerValue(item.arrivals, `${field}.arrivals`),
    departures: integerValue(item.departures, `${field}.departures`),
    total_movements: integerValue(item.total_movements, `${field}.total_movements`),
    arrival_share: ratioValue(item.arrival_share, `${field}.arrival_share`),
    departure_share: ratioValue(item.departure_share, `${field}.departure_share`),
    movements_per_hour: nonNegativeNumber(
      item.movements_per_hour,
      `${field}.movements_per_hour`
    ),
    active_aircraft: integerValue(item.active_aircraft, `${field}.active_aircraft`),
    active_routes: integerValue(item.active_routes, `${field}.active_routes`),
    observed_samples: integerValue(item.observed_samples, `${field}.observed_samples`),
    expected_samples: integerValue(item.expected_samples, `${field}.expected_samples`),
    coverage_score: ratioValue(item.coverage_score, `${field}.coverage_score`),
    freshness_score: ratioValue(item.freshness_score, `${field}.freshness_score`),
    latest_observation_at: timestampValue(
      item.latest_observation_at,
      `${field}.latest_observation_at`
    ),
    generated_at: timestampValue(item.generated_at, `${field}.generated_at`),
  }
}

function parseRankingSummary(value: unknown, field: string): AirportRankingSummary {
  const item = objectValue(value, field)
  return {
    position: integerValue(item.position, `${field}.position`),
    total_airports: integerValue(item.total_airports, `${field}.total_airports`),
    activity_score: ratioValue(item.activity_score, `${field}.activity_score`),
    data_confidence: ratioValue(item.data_confidence, `${field}.data_confidence`),
    movements_component: ratioValue(item.movements_component, `${field}.movements_component`),
    routes_component: ratioValue(item.routes_component, `${field}.routes_component`),
    observations_component: ratioValue(
      item.observations_component,
      `${field}.observations_component`
    ),
    intensity_component: ratioValue(item.intensity_component, `${field}.intensity_component`),
  }
}

function parseTrendPoint(value: unknown, field: string): AirportTrendPoint {
  const item = objectValue(value, field)
  return {
    window_start: timestampValue(item.window_start, `${field}.window_start`),
    window_end: timestampValue(item.window_end, `${field}.window_end`),
    total_movements: integerValue(item.total_movements, `${field}.total_movements`),
    movements_per_hour: nonNegativeNumber(
      item.movements_per_hour,
      `${field}.movements_per_hour`
    ),
    active_routes: integerValue(item.active_routes, `${field}.active_routes`),
    coverage_score: ratioValue(item.coverage_score, `${field}.coverage_score`),
    freshness_score: ratioValue(item.freshness_score, `${field}.freshness_score`),
  }
}

function parseLimitations(value: unknown, field: string): AirportIntelligenceLimitation[] {
  return arrayValue(value, field).map((entry, index) => {
    const item = objectValue(entry, `${field}[${index}]`)
    return {
      code: stringValue(item.code, `${field}[${index}].code`),
      message: stringValue(item.message, `${field}[${index}].message`),
    }
  })
}

function objectValue(value: unknown, field: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    invalid(`${field} must be an object.`)
  }
  return value as Record<string, unknown>
}

function arrayValue(value: unknown, field: string): unknown[] {
  if (!Array.isArray(value)) invalid(`${field} must be an array.`)
  return value
}

function stringValue(value: unknown, field: string, allowEmpty = false): string {
  if (typeof value !== 'string') invalid(`${field} must be a string.`)
  if (!allowEmpty && value.trim() === '') invalid(`${field} must not be empty.`)
  return value
}

function icaoValue(value: unknown, field: string): string {
  const normalized = stringValue(value, field).trim().toUpperCase()
  if (!/^[A-Z0-9]{4}$/.test(normalized)) invalid(`${field} must be a four-character ICAO code.`)
  return normalized
}

function finiteNumber(value: unknown, field: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    invalid(`${field} must be a finite number.`)
  }
  return value
}

function nullableFiniteNumber(value: unknown, field: string): number | null {
  return value === null ? null : finiteNumber(value, field)
}

function nonNegativeNumber(value: unknown, field: string): number {
  const parsed = finiteNumber(value, field)
  if (parsed < 0) invalid(`${field} must not be negative.`)
  return parsed
}

function boundedNumber(value: unknown, field: string, minimum: number, maximum: number): number {
  const parsed = finiteNumber(value, field)
  if (parsed < minimum || parsed > maximum) invalid(`${field} is outside its allowed range.`)
  return parsed
}

function integerValue(value: unknown, field: string, signed = false): number {
  const parsed = finiteNumber(value, field)
  if (!Number.isInteger(parsed) || (!signed && parsed < 0)) {
    invalid(`${field} must be ${signed ? 'an integer' : 'a non-negative integer'}.`)
  }
  return parsed
}

function ratioValue(value: unknown, field: string): number {
  const parsed = finiteNumber(value, field)
  if (parsed < 0 || parsed > 1) invalid(`${field} must be between zero and one.`)
  return parsed
}

function booleanValue(value: unknown, field: string): boolean {
  if (typeof value !== 'boolean') invalid(`${field} must be a boolean.`)
  return value
}

function timestampValue(value: unknown, field: string): string {
  const parsed = stringValue(value, field)
  if (Number.isNaN(Date.parse(parsed))) invalid(`${field} must be a valid timestamp.`)
  return parsed
}

function invalid(message: string): never {
  throw new APIRequestError(`The Airport Intelligence response is invalid: ${message}`)
}
